x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java

```diff
@@ -44,30 +43,27 @@ protected LogicalPlan rule(Aggregate aggregate) {
         List<Expression> newGroupings = new ArrayList<>(aggregate.groupings());
         boolean groupingChanged = false;
 
-        // start with the groupings since the aggs might duplicate it
+        // start with the groupings since the aggs might reuse/reference them
         for (int i = 0, s = newGroupings.size(); i < s; i++) {
             Expression g = newGroupings.get(i);
-            // Move the alias into an eval and replace it with its attribute.
-            // Exception: Categorize is internal to the aggregation and remains in the groupings. We move its child expression into an eval.
             if (g instanceof Alias as) {
-                if (as.child() instanceof Categorize cat) {
-                    // For Categorize grouping function, we only move the child expression into an eval
-                    if (cat.field() instanceof Attribute == false) {
+                Expression asChild = as.child();
+                // for non-evaluable grouping functions, replace their nested expressions with attributes and extract the expression out
+                // into an eval (added later below)
+                if (asChild instanceof GroupingFunction.NonEvaluatableGroupingFunction gf) {
+                    Expression newGroupingFunction = transformNonEvaluatableGroupingFunction(gf, evals);
+                    if (newGroupingFunction != gf) {
                         groupingChanged = true;
-                        var fieldAs = new Alias(as.source(), as.name(), cat.field(), null, true);
-                        var fieldAttr = fieldAs.toAttribute();
-                        evals.add(fieldAs);
-                        evalNames.put(fieldAs.name(), fieldAttr);
```

Comment:
This should not be needed, as the expression provided to the grouping function can't show up stand-alone / aliased in the aggs.